Library/Homebrew/os/linux/elf.rb

```diff
@@ -68,28 +68,30 @@ def binary_executable?
     elf_type == :executable
   end
 
-  def with_interpreter?
-    return @with_interpreter if defined? @with_interpreter
+  def interpreter
+    return @interpreter if defined? @interpreter
 
-    @with_interpreter = if binary_executable?
-      true
-    elsif dylib?
-      if HOMEBREW_PATCHELF_RB
-        begin
-          patchelf_patcher.interpreter.present?
-        rescue PatchELF::PatchError => e
-          opoo e unless e.to_s.start_with? "No interpreter found"
-          false
-        end
-      elsif which "readelf"
-        Utils.popen_read("readelf", "-l", to_path).include?(" INTERP ")
-      elsif which "file"
-        Utils.popen_read("file", "-L", "-b", to_path).include?(" interpreter ")
-      else
-        raise "Please install either readelf (from binutils) or file."
+    if HOMEBREW_PATCHELF_RB
+      begin
+        patchelf_patcher.interpreter
+      rescue PatchELF::PatchError => e
+        opoo e unless e.to_s.start_with? "No interpreter found"
+        nil
       end
+    elsif (patchelf = DevelopmentTools.locate "patchelf")
+      # the check is probably redundant, it is already done by #keg_relocate
+      # adding it as a safety precaution. And maybe for something else to use
+      # in future.
+      interp = Utils.popen_read(patchelf, "--print-interpreter", to_s, err: out).strip
+      interp.start_with?("patchelf:") ? nil : interp
+    elsif (file = DevelopmentTools.locate("file"))
+      interp = Utils.popen_read(file, "-L", "-b", to_s, err: out).strip
```

Comment:
parsing the output of `file` is troublesome.
Initially i thought of extracting text after "interpreterter " until we come across ".so,"
the problem with this approach is we can't guarantee the interpreter they picked won't have these characters in middle( e.g: `/lib/h.so,ull.so`) it may sound strange yet its a legal name to have.

Reply:
```
$ file -b /bin/ls
ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, for GNU/Linux 2.6.32, BuildID[sha1]=d0bc0fb9b3f60f72bbad3c5a1d24c9e2a1fde775, stripped
```

It's a valid concern that the interpreter may contain a comma, and thank you for raising it, but it's also pretty unlikely. The text after `, interpreter ` and before the next comma is the interpreter.

You could split the string on `, ` (comma space) and extract the element that begins with `interpreter `.

Reply:
alright, i'll leave a comment for future. so we are aware of it.

Reply:
If you split on `, ` (comma space), I think the chances of the interpreter filename containing the string `, ` (comma space) is close to nil.

Reply:
So in that case, it's probably not even worth mentioning it in a comment.

Reply:
Agreed with @sjackman here.